cgsmiles/resolve.py

```diff
@@ -185,6 +184,51 @@ def edges_from_bonding_descrpt(self):
                 order = 1
             self.molecule.add_edge(edge[0], edge[1], bonding=bonding, order=order)
 
+    def squash_atoms(self):
+        """
+        Applies the squash operator.
+        """
+        bondings = nx.get_edge_attributes(self.molecule, 'bonding')
+        squashed = False
+        for edge, bonding in bondings.items():
+            # we have a squash operator
+            if bonding[0].startswith('!'):
```

Comment:
Invert the if. If not, continue.

Even better, do the check outside of this function

Reply:
how so?

Reply:
Because if statements tend to complicate code, and this bit is already complicated.
Firstly the inversion: if you invert it, you reduce the indentation quite a bit, which simplifies reading.
As for moving the check outside the function, that will force you to separate this bit into its own function, which again helps readability. In addition, it may simplify the overall flow of code a bit, e.g.
```
for edge, bonding in bondings.items():
    if bonding[0].startswith('!'):
        squash(...)
    elif bonding[0].startswith('$'):
        stack(...)
    else:  # wut?
        raise KeyError
```

And I'll also leave https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.minors.contracted_nodes.html#networkx.algorithms.minors.contracted_nodes here

Reply:
Yes but also no, there is a logic to first to stack and then squash. Doing it this way you can have both a connect and a squash operator on the same atom. If you refactor it like suggested you need to do the additional bookkeeping of identifying what happens to the connect operator on the atom that got squashed which involves updating the fragments and doing the logic for when a fragment has more than one fragid. Not worth refactoring it just to find out that it doesn't work. 

So we keep doing first connect then squash then rebuild all fragments and done. 

The contracted nodes I should radically simplify squash though.